Adding polyfills

rorticus opened the pull request:
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

<!--
Our bots should ensure:

* [ ] All contributors have signed a CLA
* [ ] The PR passes CI testing
* [ ] Code coverage is maintained
* [ ] The PR has been reviewed and approved
-->

**Description:**

Polyfills (pepjs and intersection-observer) are now loaded in `main`. 

An AMD loader configuration mixin is now provided to make loading polyfills easier. It injects a global function that you call to mix in the polyfill/dojo2 loader configuration with your existing configuration:

```
shimAmdDependencies({
    baseUrl: '/',
    packages: [ ... ]
});
```

The intern config for shim has been updated to support this configuration utility (and all loader config has been removed).

Resolves #106

rorticus committed:
Turning on importHelpers

rorticus committed:
Adding polyfills, issue #106

rorticus commented on src/main.ts:
```diff
@@ -9,6 +11,18 @@ import * as string from './string';
+import './Symbol';
+import has from '@dojo/has/has';
+
+if (has('host-browser')) {
```
@kitsonk I couldn't just use the `!has()` string here because `pepjs` and `intersection-observer` error if loaded from node (they expect a browser). I can add the `has` checks before the `require` statements if you want?

kitsonk replied:
Let's talk about this, because while TypeScript allows it, we don't want to have a `require()` in the code...  We may need to use a dynamic `import('pepjs');` instead of the `require`.

rorticus commented on src/support/tslib.ts:
```diff
@@ -0,0 +1,41 @@
+
+import 'tslib';
+
+global.__values = function (o: any) {
```
We need to override the typescript `__values` function because it doesn't handle multibyte strings properly. This will let us use `for...of` with multibyte strings.

rorticus commented on tests/loader.ts:
```diff
@@ -0,0 +1,16 @@
+intern.registerLoader((options) => {
```
Providing our own loader lets us configure and call our shimAmdDependencies method rather than including all the configuration in the intern config. It probably makes sense to have this loader (or something like it) available for other packages to use. Thoughts?

rorticus commented on tests/loader.ts:
```diff
@@ -0,0 +1,16 @@
+	return intern.loadScript('node_modules/@dojo/loader/loader.js')
+		.then(() => intern.loadScript('./_build/src/util/amd.js'))
+		.then(() => {
+			(<any> require).config(shimAmdDependencies({
```
I tried adding the typings to `require`. As soon as I import `@dojo/interfaces/loader`, the file turns into an actual module instead of just a plain function and the whole thing breaks down in agony.

kitsonk replied:
`(require as any)` and in a test case, the lack of typings is fine, though it would be good to put a comment in the code as to why you are casting.

rorticus commented:
CI failures from typedoc

rorticus commented:
Talked to @kitsonk and @matt-gadd , we agreed that we need to,

* Separate out the browser polyfills into another module, like `@dojo/shim/browser`. This would ONLY load the polyfills.
* Change the conditional require statements into the regular pragma stuff
* Load JSDOM when running shim tests (https://github.com/dojo/test-extras/blob/master/src/support/loadJsdom.ts)

If JSDOM doesn't solve the problem ,leave out the DOM polyfills for right now.

rorticus committed:
Refactoring based on discussion

rorticus committed:
Adding intersection observer polyfill and functional tests

rorticus committed:
Turning off typedoc

rorticus committed:
Adding tests for amd shim

rorticus commented:
@kitsonk @matt-gadd Made the changes we talked about!

* Added `browsers` module for browsers to import
* Removed polyfills from `@dojo/shim/main`
* Added more tests

rorticus committed:
Fixing test..

rorticus committed:
Adding downlevel iteration

rorticus committed:
Adding import helpers back in

kitsonk commented:
I believe I am correct in assuming that this will require our tests and build to `import '@dojo/shim';` as standard practice?

kitsonk reviewed (commented):
We should really update the README with details of these changes as they are fairly significant.  Other than that, and a nit, looks 💯

rorticus commented:
@kitsonk , I'll go ahead and update the README with some friendly words. You're correct about needing to `import '@dojo/shim'`.

rorticus committed:
Updated README

rorticus committed:
Add blurb about browser module

rorticus committed:
README updates

rorticus committed:
Updates for intern4

rorticus committed:
Intern 4 updates

rorticus committed:
Fixing AMD tests

rorticus committed:
Fix for iteration problem on windows

rorticus commented:
Added a dependency on https://github.com/dojo/grunt-dojo2/pull/169 .

rorticus committed:
Patching tslib __values for webpack builds

rorticus commented:
PRs for after this is released,

- [dojo/loader](https://github.com/dojo/loader/pull/149)
- [dojo/has](https://github.com/dojo/has/pull/59)
- [dojo/core](https://github.com/dojo/core/pull/365)
- [dojo/i18n](https://github.com/dojo/i18n/pull/103)
- [dojo/widget-core](https://github.com/dojo/widget-core/pull/752)
- [dojo/routing](https://github.com/dojo/routing/pull/111)
- [dojo/test-extras](https://github.com/dojo/test-extras/pull/77)
- [dojo/cli-create-app](https://github.com/dojo/cli-create-app/pull/118)
- [dojo/cli-build-webpack](https://github.com/dojo/cli-build-webpack/pull/237)
- [dojo/widgets](https://github.com/dojo/widgets/pull/369)
- [dojo/cli-test-intern](https://github.com/dojo/cli-test-intern/pull/71)
- [dojo/cli-export-project](https://github.com/dojo/cli-export-project/pull/12)
- [dojo/cli](https://github.com/dojo/cli/pull/175)
